Cpp23

arunveeramany opened the pull request:
#### What's this Pull Request do?

Validation report:
634 tested; WLS2: 100% (except wget); WIN32: 96% (except *cloud*tmy*)

•	 Windows-Specific Includes and Macros
•	platform-specific headers and macros to handle file access checks on Windows.
•	Removed or replaced constructs that are only valid in GCC/Clang (e.g., certain inline initializations or anonymous structs).
•	Reformatted enum class and typedef struct blocks to be MSVC-friendly
•	Initialized scalar members directly in constructors or create() methods
•	Replaced std::string usage with char* for compatibility with legacy APIs and Windows-style string handling
•	Ensured buffers are null-terminated after copy operations pthreads replaced with std::thread
•        users have to install Eigen library for example, on Ubuntu: sudo apt install libeigen3-dev
•	 class gld_rlock is phased out. This is replaced with centralized SharedMutexManager. GL_ARRAY, GL_STRING, GL_ATOMIC are phased out. These macros are replaced with concrete definitions.
•	 replaced all legacy read locks (rlock, RLOCK_OBJECT etc.) with a centralized registry of mutexes.
•	revived 'class_block' and runtime compilation features only for auction and player classes
•	Use of “class”, “script”, “extern” from .glm now return error saying “deprecated”
•	Replaced double_array and complex_array with highly efficient and popular Eigen::Matrix
•	removal of “scripting” feature.
•	exec.cpp is free from “malloc” and “free” – these are now smart pointers – this is massive change.
•	Case 1: multithreading is not requested yet mandatorily used – for “commit all”   (the single threaded one is buggy).
•	Case 2: multithreading is requested and executed using threadpools (I commented out all these in exec_start, will revive later during modernization).
•	master-slave logic (multiruns?) is commented out
•	The project continues to be C++23 and C++20 compatible (not strictly compliant).
•	Cryptic pointer arithmetic to assign weather properties is replaced with human readable property names.
•	6 Infamous #defines (OBJECTHDR, OBJECTDATA, …) replaced with template functions for easy compile time debugging.
•	Commented out all the dreaded memset(this, 0, sizeof(*this)).
•	All glm objects are now inherited either directly or indirectly from gld_object allowing for pure virtual destruction.
•	Replaced a few malloc with smart allocations (e.g., weather).
•	Added overloaded version of convert_from/to_set to accept uint32 and thereby reducing invalid memory accesses.
•	Globally enforced C++20/C++23 standard across all modules (CMAKE_CXX_STANDARD 23)
•	Replaced delete macros with safer object registration via register_object<T> using std::unique_ptr
•	Deprecated gldapi across build files and CMakeLists.txt by commenting out its inclusion and linking
•	Introduced safe object tracking via allocated_objects and cleanup via atexit(cleanup_tracked_objects)
•	Avoided raw memset(this, 0, sizeof(...)) in classes like pqload to preserve class integrity
•	Refactored global_find to use std::string_view for safer string handling
•	Updated signature of property_malloc to accept std::string_view instead of char*
•	Improved destructor logic for complex_array and double_array to prevent memory leaks and dangling references
•	Replaced unsafe string size assumptions (char[1024]) with larger buffers (e.g., char[4096]) where needed
•	Added consistent whitespace trimming in multiple modules using safer strlen-based logic
•	Fixed atomic_increment implementation for GCC platforms using correct volatile cast (unsigned int*)
•	Removed redundant or obsolete calls to memory allocation and initialization (strdup, raw malloc, etc.)
•	Added dynamic registration for testable object instantiation in residential and powerflow modules
•	Updated object creation in create_triplex_meter for safer construction and better error diagnostics
•	Cleaned up dangling pointers and memory after simulation runs (free() loops in check())

#### Where should the reviewer start?
main.cpp

#### How should this be tested?
gridlabd --validate from the root
As well, a new dependency is Eigen::Matrix in place of legacy complex_array and double_array

#### Any background context you want to provide?
This is the basis for modernization and multithreading

#### What are the relevant issues?
cryptic macros, dangerous pointer arithmetic, and allocated objects not freed up

#### Screenshots (if appropriate)
#### Questions:
- [x] Does this add new dependencies?
- [x] Is there appropriate logging included?

arunveeramany committed:
readiness towards CXX 20/23

arunveeramany committed:
safety, readability

arunveeramany committed:
660 tests succeeded with added memory-safe C++23 features, and readability

arunveeramany committed:
smart pointers in exec, rescinding threadpool

kevinatkinson-pnnl commented on gldcore/load.cpp:
```diff
@@ -5661,121 +5661,121 @@ static int link_declaration(PARSER)
-		REJECT;
-}
+//////////////////////////////////////////////////////////////////////////////////
+//static int script_directive(PARSER)
```
Was this unused?

arunveeramany replied:
This was used for what is now obsolete "scripts" directive

kevinatkinson-pnnl commented on assert/complex_assert.cpp:
```diff
@@ -240,7 +240,9 @@ EXPORT SIMULATIONMODE update_complex_assert(OBJECT *obj, TIMESTAMP t0, unsigned
 	char error_output_buff[2048];
 	char datebuff[128];
-	complex_assert *da = OBJECTDATA(obj,complex_assert);
+	/*complex_assert *da = OBJECTDATA(obj,complex_assert);*/
```
Is there a reason we need to leave unused/commented code in?

arunveeramany replied:
This is for quick comparison by the reviewers

arunveeramany committed:
csv reader free of malloc

arunveeramany committed:
removed rt feature. replaced custom matrices with Eigen::Matrix

arunveeramany committed:
legacy atomic and spinlock-based read locks replaced with modern mutexes

arunveeramany committed:
Replaced complex_assert and enum_assert use of GL_ macros with concrete definitions and thereby phasing out the use of class gld_rlock

arunveeramany committed:
class gld_rlock is phased out. This is replaced with centralized SharedMutexManager. GL_ARRAY, GL_STRING, GL_ATOMIC are phased out. These macros are replaced with concrete definitions.

arunveeramany committed:
Write locks of all forms (class, functions, macros) are disabled but handled by the centralized SharedMutexManager.

shwethanidd committed:
gld api changes

arunveeramany committed:
pthreads replaced with std::thread

shwethanidd committed:
compilation fixes

kevinatkinson-pnnl committed:
Add example run for gldapi

afisher1 committed:
Merge branch 'feature/1478' of github.com:riley206-pnnl/gridlab-d into feature/1478

afisher1 committed:
troubleshooting macos-latest pull-request-workflow.yml

afisher1 committed:
more troubleshooting

afisher1 committed:
more troubleshooting

afisher1 committed:
more troubleshooting

afisher1 committed:
more troubleshooting

afisher1 committed:
need to troublshoot macos build. Not Linux.

afisher1 committed:
Helps to spell the branch correctly

afisher1 committed:
removing no longer needed macro from connection.h.
fixing failing test in windows build.

afisher1 committed:
Merge branch 'cpp23' of github.com:arunveeramany/gridlab-d into cpp23

arunveeramany committed:
vscode msvc + gcc toolchain

arunveeramany committed:
restored taxonomy feeder test cases

arunveeramany committed:
vscode for windows config

arunveeramany committed:
Merge branch 'strings_nocharbuf' into cpp23

arunveeramany committed:
restored taxonomy feeder tests

arunveeramany committed:
restored taxonomy feeder tests

arunveeramany committed:
code modernization for MSVC: taxonomy feeders 100%, reliability 100%, powerflow 98%, gldcore 97%, tape 96%, assert 80%, residential 78%, market 68%

kevinatkinson-pnnl committed:
Add basic checkpoint creation

Also be able to pass in stop and start time to run function and access individual object properties.

kevinatkinson-pnnl committed:
Add support for default times from glm

Also some cleanup, and allowing user to designate folder for checkpoint file save.

kevinatkinson-pnnl committed:
Add function to step through a sim, refactor.

kevinatkinson-pnnl committed:
Fix stepping, need to continue iterating until clock advances

kevinatkinson-pnnl committed:
Added finalize all to exit_gld to catch when only stepping

Also removed the examples for accesssing JSON values, wasn't very useful. Added small example.

afisher1 committed:
restoreing erroneously deleted autotests. deleting additional glms with depricated functionality in them.

afisher1 committed:
Merge branch 'cpp23' of github.com:arunveeramany/gridlab-d into cpp23

arunveeramany committed:
634 tested; WLS2: 100% (except wget); WIN32: 96% (except *cloud*tmy*)

afisher1 committed:
Merge branch 'cpp23' of github.com:arunveeramany/gridlab-d into cpp23

kevinatkinson-pnnl committed:
(#1540) Add basic getting / setting of properties

Also fix end of run core and module reporting

arunveeramany committed:
no mutexes in climate.h

arunveeramany committed:
Merge remote-tracking branch 'afisher1/cpp23' into cpp23

afisher1 commented on .gitignore:
```diff
@@ -79,7 +79,6 @@ gldcore/gridlabd.sh
 .cproject
 .project
 .settings/
-.vscode
```
I do not want user specific IDE files in the codebase so please keep this in there.

afisher1 commented on .vs/launch.json:
Remove from codebase please

afisher1 commented on .vscode/c_cpp_properties.json:
remove from codebase please

afisher1 commented on .vscode/launch.json:
Remove from codebase please.

afisher1 commented on CMakeLists.txt:
```diff
@@ -112,17 +149,25 @@ add_compile_options(${FORCE_ERROR} ${DISABLED_OPTIMIZATIONS})
 
 ## Set up platform specific library linking and platform identification
 if (WIN32 OR MSYS OR MINGW OR CYGWIN)
-    FetchContent_Declare(
```
This is specific to the MINGW build. Removing it will break the MINGW build.

afisher1 commented on CMakeLists.txt:
```diff
@@ -1,6 +1,43 @@
 cmake_minimum_required(VERSION 3.13)
 project(GridLAB-D)
 
+
```
Needs better separation between building with MSCV vs MSYS in Windows.

afisher1 commented on CMakePresets.json:
This seems to be a user specific file and not needed in the codebase.

afisher1 commented on gldcore/CMakeLists.txt:
```diff
@@ -1,10 +1,13 @@
 SET(GLD_MODULE_NAME gridlabd)
 
-# Load and link C/C++ files for the executeable
+set(CMAKE_BUILD_TYPE Debug) # Add this near the top if not using multi-config generator
```
Should not be hard set to Debug and set by the user.

kevinatkinson-pnnl committed:
(#1540) Revise to use nlohman json

arunveeramany committed:
user config files removed

kevinatkinson-pnnl committed:
(#1540) Remove default to debug

arunveeramany committed:
set n_proc to global_threadcount

afisher1 commented:
@arunveeramany There is behavior that has been lost. When setting --threadcount 0 this means set the threadcount to the total number of threads from the system hardware. In your current implementation this behavior seems to be no longer the case.

afisher1 commented on gldcore/validate.cpp:
```diff
@@ -739,161 +919,201 @@ int validate(int argc, char *argv[])
+
+	// Use a vector to store threads
+	std::vector<std::thread> threads;
+	n_procs = global_threadcount;
```
remove this line. It breaks functionality from lines 1025-1028

arunveeramany committed:
n_procs set to processor_count() by default

arunveeramany committed:
n_procs set to processor_count() by default; commented out n_procs=fmin(...); raise separate ticket for this

arunveeramany committed:
validate multithreading: WSL2 100%; Windows 99%

arunveeramany committed:
Kevin's JSON loader

arunveeramany committed:
resolved conflicts with json loader / gldapi PR

arunveeramany committed:
smooth exit on expected failures (_err autotests)

afisher1 commented on .github/workflows/pull-request-workflow.yml:
```diff
@@ -168,7 +168,7 @@ jobs:
         with:
           repository: 'GMLC-TDC/HELICS'
-          ref: ${{ env.helics-ref }}
+          ref: 'develop'
```
This shouldn't have been changed

afisher1 commented on _test_results/test_HVAC_balance.1.json:
Remove this personal testing file and folder.

arunveeramany committed:
gldapi fixes for MSVC

arunveeramany committed:
removed/reset some residual changes from other commits

afisher1 committed:
Merge branch 'feature/1478' into cpp23
